### PR TITLE
[Qt] simple fee bumper with user verification

### PR DESCRIPTION
**Pull request by jonasschnelli**
~~Based on #9681.~~
Add "Increase transaction fee" action to the transaction table context menu.
The user can verify the new (and old) fee before sign & commit.

bildschirmfoto 2017-02-06 um 15 31 34
bildschirmfoto 2017-02-06 um 15 31 36

### CODE REVIEW

File: src/qt/walletmodel.cpp
Comment (ryanofsky) [inline src/qt/walletmodel.cpp]: wtx && SignalsOptInRBF(*wtx)

File: src/qt/walletmodel.cpp
Comment (ryanofsky) [inline src/qt/walletmodel.cpp]: I'm surprised it's ok to be holding onto the cs_main and cs_wallet locks while waiting for user input from a confirmation dialog, since it would seem to grind the whole bitcoin node to a halt. But maybe there are other precedents for doing this.

If you wanted to avoid it, though, I think you easily could by splitting bumpfee into two functions: one which creates the new transaction, and the other which commits this.

**Comment by jonasschnelli** [comment]
Thanks for the review @ryanofsky. Yes. The current situation with holding cs_main/cs_wallet during the QMessageBox must be avoided. Using `LEAVE_CRITICAL_SECTION` and `ENTER_CRITICAL_SECTION` would be another alternative, but I'd like your idea about splitting bumpFee into a part where we create the tx and a part where we sign and commit.
The signing part could potentially be shared with `CreateTransaction` in future.

**Comment by jonasschnelli** [comment]
Rebased on top of #9681.
Fixed the `cs_main` & `cs_wallet` lock problems mentioned by @ryanofsky.

File: src/qt/walletmodel.cpp
Comment (ryanofsky) [inline src/qt/walletmodel.cpp]: Are we able to write unit tests to exercise our qt interface code? I don't know very much about qt, but googling around it seems to be possible: http://doc.qt.io/qt-5/qttestlib-tutorial3-example.html

Asking because the logic here isn't totally trivial, and this seems like something we would want automated testing for to ensure it doesn't break in the future, especially if there will be more refactorings.

File: src/qt/walletmodel.cpp
Comment (ryanofsky) [inline src/qt/walletmodel.cpp]: Seems like it would be nice for this code if CFeeBumper constructor, wallet SignTransaction method, and CFeeBumper commit could acquire the locks they need themselves. Not asking for a change, since I don't want to hold up #9697, but curious if there was something that got in the way of doing this.

**Review by ryanofsky — COMMENTED** [review]
Tested ACK f7132c3b14cf714f71a2fe59241223fc30520814.

Confirmed fee bumping, error handling, and RBF detection all work correctly. The interface is very clean and convenient, and it seems like it could be logically extended to give users control of the new fee in the future.

File: src/qt/walletmodel.cpp
Comment (ryanofsky) [inline src/qt/walletmodel.cpp]: I started writing unit tests for this change: https://github.com/ryanofsky/bitcoin/commit/9bfbdd88dd0f1e11f06403e81f39fd8df1d047c7

They are mostly complete, but have a few things I want to fix up. They also required changes to the testing framework which I'm going to split off and make a separate PR.

**Comment by laanwj** [comment]
Adding milestone for 0.15

File: src/wallet/wallet.h
Comment (jtimon) [inline src/wallet/wallet.h]: The function doesn't seem to return anything and the enum is not used anywhere it seems.
Perhaps the function could start as void in this code movement and change it afterwards?

File: src/wallet/wallet.cpp
Comment (jtimon) [inline src/wallet/wallet.cpp]: Did the previous commit compile without hash?
I though we wanted every commit to compile and pass tests (for git bisect but also for general history cleanness).

File: src/wallet/wallet.cpp
Comment (jtimon) [inline src/wallet/wallet.cpp]: This could be done when making nOldFee a parameter of this function too.

File: src/wallet/rpcwallet.cpp
Comment (jtimon) [inline src/wallet/rpcwallet.cpp]: Seems weird and unnecessary to move this to wallet and then move it back here. Perhaps SignTransaction/BumpFeePrepare/BumpFeeCommit should be created before BumpFee (ie BumpFee never being created to just remove it in the same PR) ?

**Comment by jtimon** [comment]
utACK individual commits: 8c6316038e8ebb1877d712f29171afc4bb76a159 641c99cb6f96afa166a293917f36f3cf8f28a08a (without fully verifying code movements, but look ok) f0c98b9bfa763fbef87f1d9d37853f28e9451c87
Fast review 65e2e0085a608d9f570bc911349a5953a6ca88ee
It seems to me that the history in this PR is unnecessarily complicated and not very clean (and I believe it doesn't compile commit by commit, but I haven't checked).
Needs rebase.

**Comment by ryanofsky** [comment]
@jtimon, if you are interested in reviewing this PR, the only commit worth looking at is the final one: "[Qt] simple fee bumper with user verification" (f7132c3b14cf714f71a2fe59241223fc30520814). The preceding commits which you reviewed were just pulled in from an old version of #9681, and they will disappear after #9681 is merged and this is rebased.

If you want to review the latest version of the feebumper code in #9681, that would be very helpful. #9681 is fully up to date and its commit history has been simplified.

**Comment by jonasschnelli** [comment]
Rebased (since #9681 is now merged) and overhauled.
This now also works with encrypted wallets.

**Comment by flack** [comment]
It's kind of hard to see in the confirmation dialog by how much the fee will actually increase. Could you work that somehow into your sentence, or maybe add a table like this:

Current fee |  0.00010595
------ | -------
increase |  0.00000960
**New fee** |  **0.00011555**

**Comment by jonasschnelli** [comment]
Implemented @flack's idea.
Now it looks like:

bildschirmfoto 2017-04-26 um 09 59 11

bildschirmfoto 2017-04-26 um 10 20 40

**Comment by jtimon** [comment]
Perhaps it is too much info for the confirmation, but what about showing the current feerate and the new feerate as well?

**Comment by laanwj** [comment]
@jonasschnelli  that looks good!

**Comment by laanwj** [comment]
I tested this out a bit on testnet; seems to work! one nit. I bumped a single transaction a few times, and it showed it as being sent multiple times:
untitled2

This could be scary to users, as it looks like the transaction is sent successfully multiple times, and will possibly be mined multiple times.

After restart it was better:
untitled3

File: src/qt/walletmodel.cpp
Comment (ryanofsky) [inline src/qt/walletmodel.cpp]: Here's an updated version of the tests I wrote previously: https://github.com/ryanofsky/bitcoin/commit/71c0a80fbfe95262be3b81c45474aceaccddfad3

Feel free to cherry-pick or I can submit as a separate PR.

**Review by ryanofsky — APPROVED** [review]
ACK 488ca9f9c78d7117d11e7f1f0be14d51065f7110

**Comment by jonasschnelli** [comment]
What would be the best solution regarding bump conflicts?

At the moment RPCs `listtransaction` does also list all previous tx in a bump (you can see the `walletconflicts` there).

Completely hiding the conflicted transactions (only own bump conflicts) would be a solution though I would provide to just "flagging" (color or opacity) the conflicts in mempool.
I once had a PR for that: https://github.com/bitcoin/bitcoin/pull/7826

Any ideas?

File: src/qt/sendcoinsdialog.h
Comment (jonasschnelli) [inline src/qt/sendcoinsdialog.h]: @ryanofsky: because setting `QWidget *parent = 0` is required, that would either result in swapping `QWidget *parent = 0` with `int secDelay = SEND_CONFIRM_DELAY` or passing `SEND_CONFIRM_DELAY` in `sendcoinsdialog.cpp` (as it is by now), right?
The current situation seems to be okay to me.
Or do I misunderstand something?

**Comment by jonasschnelli** [comment]
Fixed @laanwj's point.
Added a commit that ensures the table model row gets updated correctly after a bump.

File: src/qt/sendcoinsdialog.h
Comment (ryanofsky) [inline src/qt/sendcoinsdialog.h]: > Or do I misunderstand something?

No you're right. I didn't see sendcoinsdialog is also passing a this argument, so there is no way to use the default delay argument.

File: src/qt/transactionview.cpp
Comment (ryanofsky) [inline src/qt/transactionview.cpp]: In commit "Make sure we always update the table row after a bumpfee call"

I was surprised that passing false for the `updateTransaction` `showTransaction` parameter doesn't hide the transaction, though it's good that it doesn't. Do you think this parameter should be removed (separately, not part of this change)? I'm not seeing what it does other than set a temporary state that gets immediately overwritten.

**Review by ryanofsky — COMMENTED** [review]
ACK 1015ef40ea37a33faeded844e2da62cc6c66b0bf

File: src/qt/transactionview.cpp
Comment (jonasschnelli) [inline src/qt/transactionview.cpp]: Heh. Yes. This is really surprising and yes, I think we should remove it (in a clean follow up PR).

File: src/qt/walletmodel.cpp
Comment (TheBlueMatt) [inline src/qt/walletmodel.cpp]: I believe this is the first use of CFeeBumper without holding cs_wallet through the object's entire lifetime, so more thourough review of that is merited. I read through it and /think/ its OK, but it would probably be good to duplicate a few of the checks in the CFeeBumper constructor in commit, specifically the HasWalletSpend/mempool descendant count, wtx.mapValue.count("replaced_by_txid"), and GetDepthInMainChain() != 0 ones.

**Review by TheBlueMatt — COMMENTED** [review]
As for not displaying bumps, I'd advocate for just hiding them wholesale in TransactionRecord::showTransaction (unless the original, non-bumped version got confirmed).

File: src/qt/walletmodel.cpp
Comment (jonasschnelli) [inline src/qt/walletmodel.cpp]: Oh. Right. Here it misses `nTxConfirmTarget`. Will fix.

**Comment by jonasschnelli** [comment]
Added 3 commits to address @TheBlueMatt findings.

* Make sure to pass `nConfirmTarget` during Qt fee bumps (d9afe0192c9ad83594074661d55a4242f33de462)
* Re-check the conditions during feebump "commit". Because the Qt dialog may stay open for a couple of seconds (or even minutes), we need to re-check the descendants and if the tx has been mined (717495875b995c63e3e4166245e88e8def7f1e47). This still has room to optimise. We may want to timeout a bump...
*  Don't hide rows when the bump has been cancelled (c9c48ea36fab3c0c36665289cbb4038491aa1354)

**Comment by jonasschnelli** [comment]
Rebased.

File: src/wallet/feebumper.cpp
Comment (paveljanik) [inline src/wallet/feebumper.cpp]: typo "hasen't"

**Review by paveljanik — APPROVED** [review]

**Comment by paveljanik** [comment]
Minor nit: can we gray "Increase transaction fee" menu entry when on transaction that was already fee-bumped?

It has no reason to allow bumping the fee when it already was fee bumped and clicking on the menu entry will show

```
Increasing transaction fee failed
(Cannot bump transaction xyz which was already bumped by transaction abc)
```

**Comment by paveljanik** [comment]
... and now when my tx has been mined, there is no reason to show this menu entry at all (because clicking on it will tell me it is mined already).

**Comment by laanwj** [comment]
utACK a387837

I think this is ready for merging. Sure, it's not perfect yet, but it's usable and can be improved later.
It has a reasonable number of ACKs/utACKs and nits have been fixed a few times.

Anyone disagree?

**Comment by ryanofsky** [comment]
utACK a387837. I agree with merging.

Changes since my last review were just rebase on updated master and 3 new commits.

**Pull request merged**
